ls1intum/Artemis
# `Programming exercises`: Add build log analytics

### Checklist
#### General


#### Server
#### Client
- [x] I added a screenshots/screencasts of my UI changes.
#### Changes affecting Programming Exercises

### Motivation and Context
The duration of builds can vary based on different aspects, e.g. the complexity of the tests or the number of downloaded dependencies.
Instructor currently do not have an option to easily analyze what parts of the build execution take what time.

### Description
This PR analyzes build logs to extract these information, in particular:
- total job duration
- time to setup the build agent (e.g. docker pull)
- test duration
- static code analysis duration
- number of downloaded dependencies

These statistics are currently only supported for JAVA exercises, on both Jenkins and Bamboo.
The extracted statistics are then shown to the instructor aggregated:


**Note**: This PR requires updates of both the Bamboo and the Jenkins plugin: They have to send the build logs for every build, not only for failed ones as they do now.
PR for Bamboo: https://github.com/ls1intum/bamboo-server-notification-plugin/pull/27
PR for Jenkins: https://github.com/ls1intum/jenkins-server-notification-plugin/pull/20

The changes have already been installed on the testing systems and can be tested on TS1, TS2 and TS3.

### Steps for Testing
Prerequisites:
- Updated Bamboo/Jenkins plugins installed

1. Create a Java programming exercise with SCA enabled
2. Participate in the exercise multiple times
3. Average build statistics should be shown in the exercise details page 
4. Check that build logs are not persisted for successful jobs
5. Repeat steps 1-4 for Java exercises without SCA: The corresponding value should be empty detail page
6. Repeat steps 1-4 for unsupported exercises, e.g. Python: The value should be empty in the detail page

### Review Progress





#### Code Review
#### Manual Tests Bamboo
#### Manual Tests Jenkins

### Test Coverage


| Class/File | Branch | Line |
| ------------- | ---------- | ----- |
| ProgrammingExerciseGradingService | 88% (+/-0) | 99% (+/-0) |
| BambooService | 82% (+1) | 60% (+/-0) |
| JenkinsService | 81% (+1) | 77% (+/-0) |
| programming-exercise-detail.component.ts | 53.72% (+/-0) | 55.45% (+ 0.41%) |


### Screenshots
![image](https://user-images.githubusercontent.com/5084100/191667243-8ca97999-3966-4689-90ea-6d0276ffb353.png)
